erts/emulator/sys/unix/sys.c

```diff
@@ -230,6 +273,49 @@ thr_create_cleanup(void *vtcdp)
     erts_free(ERTS_ALC_T_TMP, tcdp);
 }
 
+#ifdef HAVE_LINUX_THP
+#define ALIGN_DOWN(x, a) ((void*)(((UWord)(x)) & ~((a) - 1)))
+
+static void
+enable_linux_thp_for_text(void)
+{
+    FILE *fp;
+    char *from, *to;
+    extern char etext;  /* see end(3) for details */
+
+    /*
+     * Our use of madvise(... MADV_HUGEPAGE) only makes sense when the
+     * THP behavior is set to the default of [madvise].  Check that we
+     * are in this mode and exit if we are not.
+     */
+    if (is_linux_thp_enabled() == 0)
+        return;
+
+    fp = fopen("/proc/self/maps", "r");
+    if (fp == NULL)
+        return;
+
+    /*
+     * Iterate through the current process mappings to find the text
+     * segment.  When they are found, madvise their memory region to
+     * use huge pages.
+     */
+    while (fscanf(fp, "%p-%p %*[^\n]\n", &from, &to) == 2) {
+        if (to < &etext)
+            continue;
+        if (from > &etext)
+            break;
+        if ((UWord)from % sys_large_page_size != 0)
```

Comment:
This crashes if `is_linux_thp_enabled()` returns non-zero and `/sys/kernel/mm/transparent_hugepage/hpage_pmd_size` cannot be opened, resulting in `sys_large_page_size` being zero. We've got two test rigs where this is the case, one with kernel version `4.4.74` and the other with `3.12.60`.

Perhaps `is_linux_thp_enabled` should also check `get_large_page_size() != 0`?

Reply:
Sorry, I think I let that bug creep in after I rearranged the code.

Along the lines of what you have suggested, I have added an early exit in this function if `sys_large_page_size == 0`.  That should prevent the later `% sys_large_page_size` from exploding.  This variation appealed to my aesthetic sense of having the thp enabled check just be about reading and parsing the relevant `sysfs(5)` file.  However, if you think it's better to put this check in `is_linux_thp_enabled` I am certainly okay with that too